### PR TITLE
feat: Implement market data by Chain ID

### PR DESCRIPTION
**Pull request by salimtb**
## Explanation

This pull request introduces a significant enhancement to our application's functionality by integrating a new feature that presents market data more comprehensively. This includes updates such as daily percentage changes and the fluctuation in amounts for a variety of assets. To accommodate this, we've expanded our state structure to encapsulate a broader array of data points. Below is a snippet illustrating the enriched data now available:
```
{
    "0x3845badade8e6dff049820680d1f14bd3903a5d0": {
        "id": "the-sandbox",
        "price": 0.00012453786505819405,
        "marketCap": 282288.57927502826,
        ...
        "pricePercentChange1y": -9.725813860523756
    },
    "0xdac17f958d2ee523a2206206994597c13d831ec7": {
        "id": "tether",
        "price": 0.00026894681330590396,
        "marketCap": 30060069.165317584,
        ...
        "pricePercentChange1y": -0.05851492115750438
    }
}
```

As a result, the TokenRatesController state has been meticulously updated to include this marketData, enriching our application with a more detailed and dynamic representation of asset performance. This enhancement not only broadens the scope of information we provide but also elevates the user experience by offering a granular view of market trends and asset valuations.
## Changelog

We've introduced a new marketData property to the application's state, providing a comprehensive overview of market-related information for various assets. This enhancement brings a wealth of data directly into our application.
This update significantly enriches the data available within our application, offering users detailed insights into asset performance and market trends. The marketData property is meticulously structured to ensure easy access to a wide array of information, enhancing the overall user experience by providing a granular and dynamic view of the market.

### `@metamask/package-assets-controllers`

- **ADDED**: Added `marketData`  to TokenRatesState 
- **REMOVED**: Removed `contractExchangeRates` from TokenRatesState
- **REMOVED**: Removed `contractExchangeRatesByChainId`  from TokenRatesState

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/assets-controllers/src/TokenRatesController.ts [hunk 1]
```diff
@@ -94,6 +100,8 @@ export interface TokenRatesState extends BaseState {
     Hex,
     Record<string, ContractExchangeRates>
   >;
+  contractPercentChange1d: Record<Hex, ContractExchangeRates>;
+  priceChange1d: Record<Hex, ContractExchangeRates>;
```
Comment (bergeron) [inline packages/assets-controllers/src/TokenRatesController.ts:104]: We'll want to access more of the new `includeMarketData` fields soon.  At least the following, which I think model nicest as:
```
{

    "0xChain1": {
        "0xAddress1": {
            "allTimeHigh": 0,
            "allTimeLow": 0,
            "circulatingSupply": 0,
            "dilutedMarketCap": 0,
            "marketCap": 0,
            "price": 0,
            "priceChange1d": 0,
            "pricePercentChange1d": 0,
            "totalVolume": 0
        },
        ...
    },
    ...
}
```

Rather than:
```
{
    "pricePercentChange1d": {
        "0xChain1": {
            "0xAddress1": 0,
            ...
        }
    },
    "priceChange1d": {
        "0xChain1": {
            "0xAddress1": 0,
            ...
        },
        ...
    },
    ...
}
```

Not sure if we should force old state to migrate, or stick it under a new field for back compat.

Comment (bergeron) [reply packages/assets-controllers/src/TokenRatesController.ts]: It looks like https://github.com/MetaMask/core/pull/4206/commits/295320e659f9529097766d416b4dac79b07e4a25 moved to the following structure:

```
{
    "oneDayPriceChange": {
        "0xChain1": {
            "priceChange1d": {
                "0xAddress1": 0
            },
            "pricePercentChange1d": {
                "0xAddress1": 0
            }
        }
    },
}
```

It would be more convenient to get all the market data for a token in 1 object, rather than piece it together from separate objects.  So I still prefer the schema from my original post.  You just index [chain][address] and destructure whatever fields you need.

To support the asset page, we already need to expose many of the other fields we get by requesting `includeMarketData`.  At least the ones in my original post, if you want to handle them in this PR.

Comment (salimtb) [reply packages/assets-controllers/src/TokenRatesController.ts]: emmm ok, so you want me to include all the data received from our api call and not just the percentage and amount change ? i'll implement the changes and keep you informed

**Comment by bergeron** [comment]
@salimtb  I spoke to the shared libraries devs at office hours.  Their recommendation was to make a breaking change to move to the new structure, rather than maintain backwards compatibility with the old structure.  Because having the same data in multiple places has been a source of bugs.

That would mean removing `contractExchangeRates` and `contractExchangeRatesByChainId`.  This should simplify the code in core, but does force clients to access price from the new field.   It would also require clients to run a migration.  But since prices are ephemeral, we should just be able to clear this controllers state and let the next fetch populate the new structure.

I think it'd be good to keep `marketData` at the top level, just in case we need to introduce other top level fields.  

This will probably be difficult to get on mobile, until it has upgraded to a more recent version of assets controllers.  We may want to help accelerate this vs trying a tricky patch

**Comment by salimtb** [comment]
> @salimtb I spoke to the shared libraries devs at office hours. Their recommendation was to make a breaking change to move to the new structure, rather than maintain backwards compatibility with the old structure. Because having the same data in multiple places has been a source of bugs.
> 
> That would mean removing `contractExchangeRates` and `contractExchangeRatesByChainId`. This should simplify the code in core, but does force clients to access price from the new field. It would also require clients to run a migration. But since prices are ephemeral, we should just be able to clear this controllers state and let the next fetch populate the new structure.
> 
> I think it'd be good to keep `marketData` at the top level, just in case we need to introduce other top level fields.
> 
> This will probably be difficult to get on mobile, until it has upgraded to a more recent version of assets controllers. We may want to help accelerate this vs trying a tricky patch

hey @bergeron ,
this is a much cleaner solution indeed, but requires a lot more changes on the UI extension part.
for mobile, we can also apply the changes to v18 of the assets controller, which is still feasible because we have a version of the controller for mobile.
i'll also talk to tomas and try to update the asset controller in mobile ( progressively ).
I'll make the changes on core + extension for now,  and keep you posted.
for mobile i'll discuss with the team to find the best solution

**Comment by sahar-fehri** [comment]
Hey @salimtb 👋  should the changelog explicitly indicate the removal of `contractExchangeRates`, `contractExchangeRatesByChainId` and adding the `marketData` instead in the default state? Maybe: 
**Removed** : Removed `contractExchangeRates` from TokenRatesState
**Removed** : Removed `contractExchangeRatesByChainId ` from TokenRatesState
**Added** : Added `marketData ` to TokenRatesState

File: packages/assets-controllers/src/TokenRatesController.ts [hunk 2]
```diff
@@ -73,6 +73,35 @@ export interface ContractExchangeRates {
   [address: string]: number | undefined;
```
[the remark this answers is not in the document]
Comment (sahar-fehri) [reply packages/assets-controllers/src/TokenRatesController.ts:73]: Resolving this cause we said we will update the client to remove the call for fetchTokenContractExchangeRates
and use marketData instead in future PRs

**Comment by sahar-fehri** [comment]
LGTM

**Review by sahar-fehri — APPROVED** [review]

File: packages/assets-controllers/src/TokenRatesController.ts [hunk 3]
```diff
@@ -527,15 +570,38 @@ export class TokenRatesController extends StaticIntervalPollingControllerV1<
       },
       initialResult: {},
     });
+    contractNativeInformations = tokenPricesByTokenAddress;
 
-    return Object.entries(tokenPricesByTokenAddress).reduce(
+    // fetch for native token
+    if (tokenAddresses.length === 0) {
+      const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
```
Comment (bergeron) [inline packages/assets-controllers/src/TokenRatesController.ts:577]: Can you explain why this is a separate case?  Why can't the caller explicitly pass the zero address if they want native data?

Comment (salimtb) [reply packages/assets-controllers/src/TokenRatesController.ts:577]: We'll use the zero address to access market data for the native token. This means that, regardless of whether users have imported any tokens or not, we're required to initiate at least one request using the zero address to gather market information about the native token.

File: packages/assets-controllers/src/token-prices-service/codefi-v2.ts [hunk 4]
```diff
@@ -349,19 +364,23 @@ export class CodefiTokenPricesServiceV2
     currency: SupportedCurrency;
   }): Promise<Partial<TokenPricesByTokenAddress<Hex, SupportedCurrency>>> {
     const chainIdAsNumber = hexToNumber(chainId);
+    const ZERO_ADDRESS: `0x${string}` =
```
Comment (GuiBibeau) [inline packages/assets-controllers/src/token-prices-service/codefi-v2.ts:367]: nit: worth making this a constant to share.

File: packages/assets-controllers/src/token-prices-service/codefi-v2.ts [hunk 5]
```diff
@@ -370,25 +389,52 @@ export class CodefiTokenPricesServiceV2
...
+          marketCap,
+          totalVolume,
+          low1d,
+          marketCapPercentChange1d,
+          pricePercentChange1h,
+          pricePercentChange7d,
+          pricePercentChange14d,
+          pricePercentChange30d,
+          pricePercentChange200d,
+          pricePercentChange1y,
```
Comment (GuiBibeau) [inline packages/assets-controllers/src/token-prices-service/codefi-v2.ts:437]: You can use the spread operator to avoid the verbose destructuring in `tokenData`
```
const tokenPrice: TokenPrice<Hex, SupportedCurrency> = {
  tokenAddress,
  ...tokenData,
  currency,
};
```

File: packages/assets-controllers/src/TokenRatesController.ts [hunk 6]
```diff
@@ -574,20 +640,33 @@ export class TokenRatesController extends StaticIntervalPollingControllerV1<
...
+      return {
+        marketData: {},
+      };
     }
 
-    return Object.entries(contractExchangeRates).reduce(
+    const updatedContractExchangeRates = Object.entries(
+      contractExchangeInformations.marketData,
+    ).reduce(
       (obj, [tokenAddress, tokenValue]) => {
```
Comment (GuiBibeau) [inline packages/assets-controllers/src/TokenRatesController.ts:651]: nit: `acc` or `accumulator` would be a better variable name probably

File: packages/assets-controllers/src/TokenRatesController.ts [hunk 7]
```diff
@@ -254,13 +307,14 @@ export class TokenRatesController extends StaticIntervalPollingControllerV1<
       }
     });
 
+    // eslint-disable-next-line @typescript-eslint/no-misused-promises
```
Comment (GuiBibeau) [inline packages/assets-controllers/src/TokenRatesController.ts:310]: No way to avoid this eslint disable?

File: packages/assets-controllers/src/TokenRatesController.ts [hunk 8]
```diff
@@ -240,6 +292,7 @@ export class TokenRatesController extends StaticIntervalPollingControllerV1<
...
```
Comment (GuiBibeau) [inline packages/assets-controllers/src/TokenRatesController.ts:295]: no way to avoid this eslint disable?

File: packages/assets-controllers/src/TokenRatesController.ts [hunk 9]
```diff
@@ -82,17 +111,40 @@ enum PollState {
...
+        priceChange1d: number;
+        pricePercentChange1d: number;
+        pricePercentChange1h: number;
+        pricePercentChange1y: number;
+        pricePercentChange7d: number;
+        pricePercentChange14d: number;
+        pricePercentChange30d: number;
+        pricePercentChange200d: number;
+        totalVolume: number;
+      }
```
Comment (GuiBibeau) [inline packages/assets-controllers/src/TokenRatesController.ts:146]: Is this different than the type on line 77? If yes, let's move to a single type to make future edits easier

File: packages/assets-controllers/src/TokenRatesController.ts [hunk 10]
```diff
@@ -254,13 +307,14 @@ export class TokenRatesController extends StaticIntervalPollingControllerV1<
...
 
+    // eslint-disable-next-line @typescript-eslint/no-misused-promises
     onNetworkStateChange(async ({ providerConfig }) => {
       const { chainId, ticker } = providerConfig;
       if (
         this.config.chainId !== chainId ||
         this.config.nativeCurrency !== ticker
       ) {
-        this.update({ contractExchangeRates: {} });
+        this.update({ marketData: {} });
```
Comment (GuiBibeau) [inline packages/assets-controllers/src/TokenRatesController.ts:317]: if the default state updates we could miss having to update it here.,

```suggestion
        this.update(this.defaultState);
```

File: packages/assets-controllers/src/TokenRatesController.ts [hunk 11]
```diff
@@ -447,14 +483,21 @@ export class TokenRatesController extends StaticIntervalPollingControllerV1<
...
-        };
-      }, {});
+      return tokenAddresses.reduce(
+        (obj, tokenAddress) => {
+          obj.marketData = {
+            ...obj.marketData,
+            [tokenAddress]: undefined,
+          };
+
+          return obj;
```
Comment (GuiBibeau) [inline packages/assets-controllers/src/TokenRatesController.ts:495]: I would look into using the delete keyword instead of setting undefined explicitely. We should avoid setting undefined explicitely.

proposal:

```
 const newMarketData = { ...obj.marketData };
  delete newMarketData[tokenAddress];
  return {
    ...obj,
    marketData: newMarketData,
  };
```

File: packages/assets-controllers/src/token-prices-service/codefi-v2.ts [hunk 12]
```diff
@@ -349,19 +364,23 @@ export class CodefiTokenPricesServiceV2
...
-    > = await this.#tokenPricePolicy.execute(() =>
-      handleFetch(url, { headers: { 'Cache-Control': 'no-cache' } }),
-    );
+    const pricesByCurrencyByTokenAddress: AddressCryptoDataMap =
+      await this.#tokenPricePolicy.execute(() =>
+        handleFetch(url, { headers: { 'Cache-Control': 'no-cache' } }),
+      );
 
-    return tokenAddresses.reduce(
+    return [ZERO_ADDRESS, ...tokenAddresses].reduce(
```
Comment (mcmire) [inline packages/assets-controllers/src/token-prices-service/codefi-v2.ts:383]: Why are we passing this now?

File: packages/assets-controllers/src/token-prices-service/abstract-token-prices-service.ts
Comment (mcmire) [inline packages/assets-controllers/src/token-prices-service/abstract-token-prices-service.ts:1]: Now that the API no longer returns a simple number but an object, does `TokenPrice` have the right name? It seems like using it in context would feel a bit off:

```
tokenPrice.price
tokenPrice.totalVolume
tokenPrice.marketCap
```

Collectively, what do all of these properties have in common? Well, they're all associated with the token. So, you could call it a `Token` instead. Then using it would look something like:

```
token.address
token.price
token.currency
token.totalVolume
token.marketCap
```

File: packages/assets-controllers/src/token-prices-service/codefi-v2.ts [hunk 13]
```diff
@@ -370,25 +389,52 @@ export class CodefiTokenPricesServiceV2
...
         const lowercasedTokenAddress =
           tokenAddress.toLowerCase() as Lowercase<Hex>;
-        const lowercasedCurrency =
-          currency.toLowerCase() as Lowercase<SupportedCurrency>;
 
-        const price =
-          pricesByCurrencyByTokenAddress[lowercasedTokenAddress]?.[
-            lowercasedCurrency
-          ];
+        const tokenData =
```
Comment (mcmire) [inline packages/assets-controllers/src/token-prices-service/codefi-v2.ts:393]: It appears that if `pricesByCurrencyByTokenAddress[lowercasedTokenAddress]` is `undefined`, this variable becomes `{}`. This means that any of the properties that you are destructuring below could also be `undefined`. 

At first glance, it appears that you're adding these properties to `tokenPrice` to add `tokenPrice` to the final object later. But then, there's a check below that prevent `tokenPrice` from being added to the final object (`tokenPrice.value !== undefined`). That's great! But it's a pretty roundabout way of skipping all of this code if `pricesByCurrencyByTokenAddress[lowercasedTokenAddress]` is `undefined`. So, instead of `|| {}`, what are your thoughts on doing an early return here? Something like:

``` typescript
        const tokenData =
          pricesByCurrencyByTokenAddress[lowercasedTokenAddress];

        if (tokenData === undefined) {
          return obj;
        }
```

And then you can say:

``` typescript
        const { price, ...rest } = tokenData;

        const tokenPrice: TokenPrice<Hex, SupportedCurrency> = {
          tokenAddress,
          value: price,
          currency,
          ...rest
        };

        return {
          ...obj,
          [tokenAddress]: tokenPrice
        };
```

This should also save you from having to mark these properties as optional in the `TokenPrice` type.

(Although, see note elsewhere for what to call `tokenPrice` here. Perhaps something like `tokenMarketData` would be more apt, and then this could simply become:)

``` typescript
        const marketData =
          pricesByCurrencyByTokenAddress[lowercasedTokenAddress];

        if (marketData === undefined) {
          return obj;
        }

        const token: Token<Hex, SupportedCurrency> = {
          address,
          currency,
          marketData,
        };
```

File: packages/assets-controllers/src/token-prices-service/abstract-token-prices-service.ts [hunk 14]
```diff
@@ -7,6 +7,24 @@ export type TokenPrice<TokenAddress extends Hex, Currency extends string> = {
   tokenAddress: TokenAddress;
   value: number;
   currency: Currency;
+  allTimeHigh?: number;
```
Comment (mcmire) [inline packages/assets-controllers/src/token-prices-service/abstract-token-prices-service.ts:10]: Do all of these properties need to be optional? It seems like this would add unnecessary logic to code that wants to make use of them, as they'd constantly need to check for `undefined`. See my notes in CodefiV2 on how we could avoid this.

File: packages/assets-controllers/src/token-prices-service/codefi-v2.ts [hunk 15]
```diff
@@ -370,25 +389,52 @@ export class CodefiTokenPricesServiceV2
         // to keep track of them and make sure we return the original versions.
         const lowercasedTokenAddress =
           tokenAddress.toLowerCase() as Lowercase<Hex>;
-        const lowercasedCurrency =
```
Comment (mcmire) [inline packages/assets-controllers/src/token-prices-service/codefi-v2.ts:373]: Is this no longer valid?

File: packages/assets-controllers/src/token-prices-service/codefi-v2.ts [hunk 16]
```diff
@@ -257,6 +249,29 @@ const DEFAULT_TOKEN_PRICE_MAX_CONSECUTIVE_FAILURES =
 
 const DEFAULT_DEGRADED_THRESHOLD = 5_000;
 
+type CryptoData = {
+  allTimeHigh: number;
+  allTimeLow: number;
+  circulatingSupply: number;
```
Comment (mcmire) [inline packages/assets-controllers/src/token-prices-service/codefi-v2.ts:255]: Since the parameter we pass to the API is called `includeMarketData`, should this be called `MarketData` instead of `CryptoData`?

Also, what are your thoughts on adding JSDoc for this type. In particularly, it'd be nice to explain what each property does.

Something like:

``` suggestion
/**
 * The shape of the data that the /spot-prices endpoint returns.
 */
type MarketData = {
  /**
   * The all-time highest price of the token.
   */
  allTimeHigh: number;
  /**
   * The all-time lowest price of the token.
   */
  allTimeLow: number;
  /**
   * The number of tokens in circulation.
   */
  circulatingSupply: number;
```

File: packages/assets-controllers/src/token-prices-service/codefi-v2.ts [hunk 17]
```diff
@@ -257,6 +249,29 @@ const DEFAULT_TOKEN_PRICE_MAX_CONSECUTIVE_FAILURES =
...
+  pricePercentChange1h: number;
+  pricePercentChange1y: number;
+  pricePercentChange7d: number;
+  pricePercentChange14d: number;
+  pricePercentChange30d: number;
+  pricePercentChange200d: number;
+  totalVolume: number;
+};
+
+type AddressCryptoDataMap = { [address: Hex]: CryptoData };
```
Comment (mcmire) [inline packages/assets-controllers/src/token-prices-service/codefi-v2.ts:273]: Since the addresses here are _token_ addresses and not account addresses, should we be more clear? Also, do we need to include the word "Map" in the name of the type? This fact can be gleaned by looking at the definition, and we don't typically do this for other data types such as lists, sets, etc.

What are your thoughts on:

``` suggestion
type CryptoDataByTokenAddress = { [address: Hex]: CryptoData };
```

or

``` suggestion
type MarketDataByTokenAddress = { [address: Hex]: MarketData };
```

File: packages/assets-controllers/src/token-prices-service/codefi-v2.ts [hunk 18]
```diff
@@ -349,19 +364,23 @@ export class CodefiTokenPricesServiceV2
...
     url.searchParams.append('vsCurrency', currency);
+    url.searchParams.append('includeMarketData', 'true');
 
-    const pricesByCurrencyByTokenAddress: SpotPricesEndpointData<
-      Lowercase<Hex>,
-      Lowercase<SupportedCurrency>
-    > = await this.#tokenPricePolicy.execute(() =>
-      handleFetch(url, { headers: { 'Cache-Control': 'no-cache' } }),
-    );
+    const pricesByCurrencyByTokenAddress: AddressCryptoDataMap =
```
Comment (mcmire) [inline packages/assets-controllers/src/token-prices-service/codefi-v2.ts:378]: Is this an accurate name for this data? Perhaps we should follow the name of the type here, so, something like this (or whatever we settle on as the name of the type):

``` suggestion
    const addressCryptoDataMap: AddressCryptoDataMap =

File: packages/assets-controllers/src/token-prices-service/codefi-v2.test.ts [hunk 19]
```diff
@@ -40,20 +117,93 @@ describe('CodefiTokenPricesServiceV2', () => {
         });
 
       expect(pricedTokensByAddress).toStrictEqual({
```
Comment (mcmire) [inline packages/assets-controllers/src/token-prices-service/codefi-v2.test.ts:119]: It might be good to rename these variables to suit the name of the type here once we've settled on that. So, perhaps `tokensByAddress` or something.

**Review by mcmire — COMMENTED** [review]
Sorry for the delay in reviewing. I made some suggestions below, mostly having to do with naming and updating of JSDocs.

File: packages/assets-controllers/src/TokenRatesController.ts (hunk 3 above)
Comment (mcmire) [reply packages/assets-controllers/src/TokenRatesController.ts:577]: Can we explain that somewhere here, perhaps in a comment? I also did not understand why that was necessary just by looking at this.

Besides this, it seems that we are already passing the zero address to the API directly in `fetchTokenPrices`. So, do we have to do this here?

File: packages/assets-controllers/src/TokenRatesController.ts [hunk 20]
```diff
@@ -527,15 +570,38 @@ export class TokenRatesController extends StaticIntervalPollingControllerV1<
...
       initialResult: {},
     });
+    contractNativeInformations = tokenPricesByTokenAddress;
 
-    return Object.entries(tokenPricesByTokenAddress).reduce(
+    // fetch for native token
+    if (tokenAddresses.length === 0) {
+      const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
+
+      const contractNativeInformationsNative =
```
[the remark this answers is not in the document]
Comment (salimtb) [reply packages/assets-controllers/src/TokenRatesController.ts:579]: When a user has imported ERC20 tokens, the API call includes the zero address along with the addresses of all those tokens. Conversely, if the user hasn't imported any tokens, the API is queried solely with the zero address to retrieve market data for the native token.

File: packages/assets-controllers/src/token-prices-service/codefi-v2.ts (hunk 12 above)
Comment (salimtb) [reply packages/assets-controllers/src/token-prices-service/codefi-v2.ts:383]: same explanation , when a user has imported ERC20 tokens, the API call includes the zero address along with the addresses of all those tokens to get the market data of the native token

File: packages/assets-controllers/src/TokenRatesController.ts (hunk 3 above)
Comment (salimtb) [reply packages/assets-controllers/src/TokenRatesController.ts]: @mcmire is this means that we'll make two calls api ? one for native chips and the other for erc20s ?

File: packages/assets-controllers/src/token-prices-service/codefi-v2.ts (hunk 15 above)
Comment (salimtb) [reply packages/assets-controllers/src/token-prices-service/codefi-v2.ts:373]: yes, we don't use it anymore

**Comment by bergeron** [comment]
Be sure to update PR description/title to reflect the broader market data being added.  Since it's more than just the daily % change now.

File: packages/assets-controllers/src/token-prices-service/codefi-v2.ts [hunk 21]
```diff
@@ -155,6 +147,12 @@ export const SUPPORTED_CURRENCIES = [
   'sats',
 ] as const;
 
+/**
+ * The zero address
```
Comment (mcmire) [inline packages/assets-controllers/src/token-prices-service/codefi-v2.ts:151]: Remember that we can write whatever we want for a JSDoc block. Usually, we go into more detail about the thing that we are writing the JSDocs for, but in this case I think it is fairly obvious (both from the name and also the value) that this is a zero address. So instead of saying what it is, what are your thoughts on explaining _why_ it's needed? What purpose does this serve and _how_ is it used?

File: packages/assets-controllers/src/TokenRatesController.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (mcmire) [reply packages/assets-controllers/src/TokenRatesController.ts:547]: I gotcha.

Aren't are already passing a zero address when we make the `fetchTokenPrices` call above? Why do we have to do this separately?

File: packages/assets-controllers/src/TokenRatesController.ts [hunk 22]
```diff
@@ -82,18 +111,13 @@ enum PollState {
  * @type TokenRatesState
  *
  * Token rates controller state
- * @property contractExchangeRates - Hash of token contract addresses to exchange rates (single globally selected chain, will be deprecated soon)
- * @property contractExchangeRatesByChainId - Hash of token contract addresses to exchange rates keyed by chain ID and native currency (ticker)
+ * @property marketData - Hash of token contract addresses to market data
```
Comment (mcmire) [inline packages/assets-controllers/src/TokenRatesController.ts:114]: Hmm, this is a bit more than that. It seems like this collection is still keyed by chain ID, but then after that, it is keyed by token contract address. What about:

``` suggestion
 * @property marketData - The market data for tokens on a chain.
```

or, if you want to be more specific:

``` suggestion
 * @property marketData - Market data for tokens, keyed by chain ID and then token contract address.
```

File: packages/assets-controllers/src/TokenRatesController.ts [hunk 23]
```diff
@@ -73,6 +74,34 @@ export interface ContractExchangeRates {
...
+  pricePercentChange1h: number;
+  pricePercentChange1y: number;
+  pricePercentChange7d: number;
+  pricePercentChange14d: number;
+  pricePercentChange30d: number;
+  pricePercentChange200d: number;
+  totalVolume: number;
+};
+
+export type ContractMarketData = {
```
Comment (mcmire) [inline packages/assets-controllers/src/TokenRatesController.ts:101]: We have a type called "market data" that has a property called "market data". This seems a bit misrepresentative.

Besides this, I guess I'm curious why this type is needed since it only has one property? It seems like in all the places where we return this object we could return a `Record<Hex, MarketDataDetails>` instead.

If that's the case, what if we rename `MarketDataDetails` to `TokenMarketData`? Then perhaps we don't need this type anymore — anywhere we need it, we can say `Record<Hex, TokenMarketData>`, and we can understand it to mean a collection of `TokenMarketData`s without having to give it a name.

What do you think?

File: packages/assets-controllers/src/TokenRatesController.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (salimtb) [reply packages/assets-controllers/src/TokenRatesController.ts:540]: fixed , thnx for the comment

File: packages/assets-controllers/src/TokenRatesController.ts [hunk 24]
```diff
@@ -527,13 +533,32 @@ export class TokenRatesController extends StaticIntervalPollingControllerV1<
       },
       initialResult: {},
     });
+    contractNativeInformations = tokenPricesByTokenAddress;
 
-    return Object.entries(tokenPricesByTokenAddress).reduce(
-      (obj, [tokenAddress, tokenPrice]) => {
-        return {
+    // fetch for native token
+    if (tokenAddresses.length === 0) {
```
Comment (mcmire) [inline packages/assets-controllers/src/TokenRatesController.ts:539]: Sorry, but I'm still not sure why we're doing this.

The `fetchTokenPrices` method always passes a zero address as the first argument. Therefore, it seems that even if `tokenAddresses` is empty, `tokenPricesByTokenAddress` will always contain a property which is the zero address. Therefore, it seems that we already have the functionality we need, and we don't need to do this explicitly. I would expect that this would be enough:

``` typescript
const tokenPricesByTokenAddress = await reduceInBatchesSerially<
  Hex,
  Awaited<ReturnType<AbstractTokenPricesService['fetchTokenPrices']>>
>({
  values: [...tokenAddresses].sort(),
  batchSize: TOKEN_PRICES_BATCH_SIZE,
  eachBatch: async (allTokenPricesByTokenAddress, batch) => {
    const tokenPricesByTokenAddressForBatch =
      await this.#tokenPricesService.fetchTokenPrices({
        tokenAddresses: batch,
        chainId,
        currency: nativeCurrency,
      });
    return {
      ...allTokenPricesByTokenAddress,
      ...tokenPricesByTokenAddressForBatch,
    };
  },
  initialResult: {},
});

return Object.entries(tokenPricesByTokenAddress).reduce(
  (obj, [tokenAddress, token]) => {
    return {
      ...obj,
      [tokenAddress.toLowerCase()]: token,
    };
  },
  {},
);
```

Is this true?

One other thing that's interesting to me is that in this second call, we are passing `this.config.chainId`. If we do truly want to keep this code, I feel that we should use `chainId` instead. `this.config.chainId` stores the globally selected chain ID, but `chainId` may be a different chain than the globally selected version (side note: this controller uses two different ways of polling, and eventually `updateExchangeRates` and `this.config.chainId` will go away; this is because dapps may want to connect to a different chain than the user has selected in MetaMask).

Comment (salimtb) [reply packages/assets-controllers/src/TokenRatesController.ts:539]: hey @mcmire ,
maybe my explanation wasn't clear,
to elucidate why the fetch operation for the native token is maintained even when `tokenAddresses` is an empty array, let's delve into the specifics of your proposal and clarify why it falls short in scenarios where a user hasn't imported any tokens and only holds the native token.

The `reduceInBatchesSerially` function is designed to process a collection of values in batches, determined by the `values` parameter. This batching mechanism is crucial for handling large datasets efficiently by breaking them down into manageable chunks.

In the scenario at hand, we encounter a situation where `tokenAddresses` is an empty array. This directly impacts the `reduceInBatchesSerially` function because it relies on the values array to partition the data into batches. With `tokenAddresses` being empty, the function identifies zero batches to process. Consequently, it bypasses the loop entirely ( check the function `reduceInBatchesSerially` ), leading to the immediate return of the `initialResult`, which, in this context, is an empty object. As a result, our intended API call to fetch token prices is never executed.

This behavior underscores the necessity of retaining the fetch operation for the native token outside of the `reduceInBatchesSerially` function when `tokenAddresses.length === 0`. By doing so, we ensure that essential market data for the native token is still retrieved and processed, even in the absence of any imported tokens. This approach guarantees that users solely holding the native token are not overlooked and that their token data is accurately reflected in the application.

In summary, when the list of token addresses is empty, our function `reduceInBatchesSerially` skips processing and returns an empty result. This is why we must ensure that data for the native token is fetched separately, guaranteeing that users holding only the native token still receive relevant market information.

Comment (salimtb) [reply packages/assets-controllers/src/TokenRatesController.ts:539]: @mcmire fixed for the `chainId`

Comment (mcmire) [reply packages/assets-controllers/src/TokenRatesController.ts:542]: Ah, gotcha. Thanks for the explanation. It sounds like a better fix would be to move the batching code into the service object. This way we can ensure that token addresses contains the zero address first before attempting to iterate over them. This is something that I had meant to do when I introduced that code but didn't get around to it.

I guess we don't have to do that here, though, we can do that in another PR.

**Review by mcmire — APPROVED** [review]
LGTM, thanks for your changes.
